### PR TITLE
Typescript Declaration

### PR DESCRIPTION
**Pull request by uny**
Revised version of https://github.com/howdyai/botkit/pull/494.
I checked that this declaration file passed the examples on https://github.com/howdyai/botkit/tree/master/examples.

### CODE REVIEW

File: lib/Botkit.d.ts [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+    readonly studio: Studio<S, M, B>;
+    readonly log: {
+      (...params: any[]): void;
+    }
+    hears(keywords: string | string[] | RegExp | RegExp[], events: string | string[], cb: HearsCallback<S, M, B>): this;
+    hears(keywords: string | string[] | RegExp | RegExp[], events: string | string[], middleware_or_cb: HearsFunction<M>, cb: HearsCallback<S, M, B>): this;
+    on(event: string, cb: HearsCallback<S, M, B>): this;
+    setupWebserver(port: number | string, cb: (err: Error, webserver: any) => void): this;
+    spawn(config?: S, cb?: (worker: B) => void): B;
+    startTicking(): void;
```
Comment (Naktibalda) [inline lib/Botkit.d.ts]: middleware property is not declared.
https://github.com/howdyai/botkit/blob/aeb1b025f1fd8f69879ebe33096ad3bd3eb63456/lib/CoreBot.js#L42

Comment (Naktibalda) [inline lib/Botkit.d.ts]: middleware property is not declared in Controller.
https://github.com/howdyai/botkit/blob/aeb1b025f1fd8f69879ebe33096ad3bd3eb63456/lib/CoreBot.js#L42

Comment (uny) [inline lib/Botkit.d.ts]: Thank you, I did in https://github.com/howdyai/botkit/pull/953/commits/fbe3b320db530b03a025cef65601193601f17cbb.

**Comment by Naktibalda** [comment]
Thanks, I don't get any type errors anymore, but I get a runtime error `TypeError: Cannot read property 'consolebot' of undefined`.

Compiled code looks like this:
```js
var botkit_1 = require("botkit");
var bot_options = {
    json_file_store: __dirname + '/../.data/db/'
};
// Create the Botkit controller, which controls all instances of the bot.
var controller = botkit_1.default.consolebot(bot_options);
```

File: lib/Botkit.d.ts [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+    name?: string;
+  }
+  type ConversationCallback<M extends Message> = ((message: M, convo: Conversation<M>) => void) | ({ pattern?: string | RegExp; default?: boolean; callback: (message: M, convo: Conversation<M>) => void; }[]);
+  type ConversationStatusType = 'completed' | 'active' | 'stopped' | 'timeout' | 'ending' | 'inactive';
+  type HearsCallback<S, M extends Message, B extends Bot<S, M>> = (bot: B, message: M) => void;
+  type HearsFunction<M extends Message> = (tests: string | string[] | RegExp | RegExp[], message: M) => boolean;
+  type SlackWebAPIMethod = (data: any, cb: (err: Error, response: any) => void) => void;
+}
+
+export = botkit;
```
Comment (uny) [inline lib/Botkit.d.ts]: @Naktibalda Thank you for your review.
There is my mistake.
https://github.com/howdyai/botkit/pull/953/commits/5951c740ed0448c86f1d62643b69472a785767d2
Please retry by importing as below:
```typescript
import Botkit = require('botkit');
```

**Comment by Naktibalda** [comment]
Thanks, it works now.

Could you add at least one example file converted to Typescript?

I was going to ask you to add dtslint to CI configuration, but I just noticed that botkit has no CI.

**Comment by uny** [comment]
@Naktibalda 
>Thanks, it works now.
>Could you add at least one example file converted to Typescript?
>I was going to ask you to add dtslint to CI configuration, but I just noticed that botkit has no CI.

Sure.
Is this okay? https://github.com/howdyai/botkit/pull/953/commits/b9800bca650bd82c58b33c5b506ae011a7412bb4

**Comment by Naktibalda** [comment]
I expected to see more types, but it is okay.

@benbrown I think that this PR is good to merge.

**Comment by aaronsky** [comment]
nice work!

**Pull request merged**